src/crimson/os/seastore/journal/circular_bounded_journal.cc

```diff
@@ -262,111 +366,50 @@ Journal::replay_ret CircularBoundedJournal::replay(
     header = head;
     DEBUG("header : {}", header);
     initialized = true;
-    written_to.segment_seq = NULL_SEG_SEQ;
-    auto tail = get_dirty_tail() <= get_alloc_tail() ?
-      get_dirty_tail() : get_alloc_tail();
-    set_written_to(tail);
     return seastar::do_with(
-      bool(false),
-      rbm_abs_addr(get_rbm_addr(tail)),
       std::move(delta_handler),
-      segment_seq_t(NULL_SEG_SEQ),
-      [this, FNAME](auto &is_rolled, auto &cursor_addr, auto &d_handler, auto &expected_seq) {
-      return crimson::repeat(
-	[this, &is_rolled, &cursor_addr, &d_handler, &expected_seq, FNAME]() mutable
-	-> replay_ertr::future<seastar::stop_iteration> {
-	paddr_t record_paddr = convert_abs_addr_to_paddr(
-	  cursor_addr,
-	  get_device_id());
-	return read_record(record_paddr, expected_seq
-	).safe_then([this, &is_rolled, &cursor_addr, &d_handler, &expected_seq, FNAME](auto ret)
-	    -> replay_ertr::future<seastar::stop_iteration> {
-	  if (!ret.has_value()) {
-	    if (expected_seq == NULL_SEG_SEQ || is_rolled) {
-	      DEBUG("no more records, stop replaying");
-	      return replay_ertr::make_ready_future<
-		seastar::stop_iteration>(seastar::stop_iteration::yes);
-	    } else {
-	      cursor_addr = get_records_start();
-	      ++expected_seq;
-	      is_rolled = true;
-	      return replay_ertr::make_ready_future<
-		seastar::stop_iteration>(seastar::stop_iteration::no);
+      std::map<paddr_t, journal_seq_t>(),
+      [this](auto &d_handler, auto &map) {
+      auto add_invalid_records = [&map](
+        const auto &offsets,
+        const auto &e,
+	sea_time_point modify_time)
+      {
+	if (e.type == extent_types_t::ALLOC_INFO) {
+	  alloc_delta_t alloc_delta;
+	  decode(alloc_delta, e.bl);
+	  if (alloc_delta.op == alloc_delta_t::op_types_t::CLEAR) {
+	    for (auto &alloc_blk : alloc_delta.alloc_blk_ranges) {
+	      map[alloc_blk.paddr] = offsets.write_result.start_seq;
 	    }
 	  }
-	  auto [r_header, bl] = *ret;
-	  bufferlist mdbuf;
-	  mdbuf.substr_of(bl, 0, r_header.mdlength);
-	  paddr_t record_block_base = paddr_t::make_blk_paddr(
-	    get_device_id(), cursor_addr + r_header.mdlength);
-	  auto maybe_record_deltas_list = try_decode_deltas(
-	    r_header, mdbuf, record_block_base);
-	  if (!maybe_record_deltas_list) {
-	    // This should be impossible, we did check the crc on the mdbuf
-	    ERROR("unable to decode deltas for record {} at {}",
-		  r_header, record_block_base);
-	    return crimson::ct_error::input_output_error::make();
+	}
+	return replay_ertr::make_ready_future<bool>(true);
+      };
+      written_to.segment_seq = NULL_SEG_SEQ;
+      auto tail = get_dirty_tail() <= get_alloc_tail() ?
+	get_dirty_tail() : get_alloc_tail();
+      set_written_to(tail);
+      return for_each_records(std::move(add_invalid_records), tail
```

Comment:
This first pass is supposed to be shared between both circular and segment journals in order to fully decouple `Cache` and backend specific implementations, e.g. `Cache::segment_provider`.

This seems to be challenging with the current code as the cbj and segment_journal don't share necessary implementations. I think this will lead to further unreusable code between the two journals.

The circular journal seems to be identical to the segment journal at the level of the record/delta format, and it will need to address the same issues that the segment journal is already addressed. For example record-batching, concurrent record submission, write-size/fullness controls, and the according scanning-record implementations.

A possible direction might be to extend `RecordSubmitter` to be reused by circular journal, then reuse the scanning part `SegmentManagerGroup::scan_valid_records()`. After record submission and scanning are generalized in the `Journal` component, reusing the two-pass replaying would be simple.

Reply:
Agree. I'll add this (extend RecordSubmitter) to my to-do-list.

So, there are thee items in my to-do-list.
* RBM step 2
* RBMCleaner---implement TODO, multiple device, and other functions
* Extend RecordSubmitter

Make sense?

Reply:
Can you elaborate more about the RBM step 2?

btw, the RBM OOL PR will still need @athanatos's review, he's on PTO this week.

Reply:
https://docs.google.com/document/d/1O9h1dMrnTAqMV7_tImoxTovgC8v1nyetXLbKsKYrxfc/edit

As we discussed, step 2 would contain A. metadata (with regard to this term, please refer to the top line of the above document) area in RBM and B. apply method (apply records from CBJ to RBM).

Reply:
@cyx1231st I added comments to the above document.